cardano-crypto-class/src/Cardano/Crypto/Libsodium/Memory/Internal.hs

```diff
@@ -130,3 +183,89 @@ sodiumFree ptr size = do
     errno <- getErrno
     ioException $ errnoToIOError "c_sodium_munlock" errno Nothing Nothing
   c_sodium_free ptr
+
+zeroMem :: MonadST m => Ptr a -> CSize -> m ()
+zeroMem ptr size = unsafeIOToMonadST . void $ c_memset (castPtr ptr) 0 size
+
+copyMem :: MonadST m => Ptr a -> Ptr a -> CSize -> m ()
+copyMem dst src size = unsafeIOToMonadST . void $ c_memcpy (castPtr dst) (castPtr src) size
+
+allocaBytes :: Int -> (Ptr a -> ST s b) -> ST s b
+allocaBytes size f =
+  unsafeIOToST $ Foreign.allocaBytes size (unsafeSTToIO . f)
+
+useByteStringAsCStringLen :: ByteString -> (CStringLen -> ST s a) -> ST s a
+useByteStringAsCStringLen bs f =
+  allocaBytes (BS.length bs + 1) $ \buf -> do
+    len <- unsafeIOToST $ BS.unsafeUseAsCStringLen bs $ \(ptr, len) ->
+      len <$ copyMem buf ptr (fromIntegral len)
+    f (buf, len)
+
+packByteStringCStringLen :: MonadST m => CStringLen -> m ByteString
+packByteStringCStringLen (ptr, len) =
+  withLiftST $ \lift -> lift . unsafeIOToST $ BS.packCStringLen (ptr, len)
+
+newtype MLockedAllocator m =
+  MLockedAllocator
+    { mlAllocate :: forall a. CSize -> m (MLockedForeignPtr a)
+    }
+
+mlockedAllocaSized :: forall m n b. (MonadST m, MonadThrow m, KnownNat n) => (SizedPtr n -> m b) -> m b
+mlockedAllocaSized = mlockedAllocaSizedWith mlockedMalloc
+
+mlockedAllocaSizedWith ::
+     forall m n b. (MonadST m, MonadThrow m, KnownNat n)
+  => MLockedAllocator m
+  -> (SizedPtr n -> m b)
+  -> m b
+mlockedAllocaSizedWith allocator k = mlockedAllocaWith allocator size (k . SizedPtr) where
+    size :: CSize
+    size = fromInteger (natVal (Proxy @n))
+
+mlockedAllocForeignPtrBytes :: MonadST m => CSize -> CSize -> m (MLockedForeignPtr a)
+mlockedAllocForeignPtrBytes = mlockedAllocForeignPtrBytesWith mlockedMalloc
+
+mlockedAllocForeignPtrBytesWith :: MLockedAllocator m -> CSize -> CSize -> m (MLockedForeignPtr a)
+mlockedAllocForeignPtrBytesWith allocator size align = do
+  mlAllocate allocator size'
+  where
+    size' :: CSize
+    size'
+        | m == 0    = size
```

Comment:
Protect against division by zero. Alternatively we could raise an actual `error`, since zero alignment doesn't make sense
```suggestion
        | m == 0     = size
        | align == 0 = 0
```